File: Beakjoon/src/Gold/Lv5/boj_1374.java
package Gold.Lv5;

import java.io.BufferedWriter;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.PriorityQueue;

public class boj_1374 { // 강의실
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        int[][] arr = new int[N][3];

        for (int i = 0; i < N; i++) {
            String[] input = br.readLine().split(" ");

            arr[i][0] = Integer.parseInt(input[0]);
            arr[i][1] = Integer.parseInt(input[1]);
            arr[i][2] = Integer.parseInt(input[2]);
        }

        Arrays.sort(arr, (o1, o2) -> {
            if (o1[1] == o2[1]) return o1[2] - o2[2];
            return o1[1] - o2[1];
        });

        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int i = 0; i < N; i++) {
            if(pq.isEmpty()) {
                pq.add(arr[i][2]);
            } else {
                int end_time = pq.peek();
                if(arr[i][1] < end_time) {
                    pq.add(arr[i][2]);
                } else {
                    pq.poll();
                    pq.add(arr[i][2]);
                }
            }
        }

        bw.write(pq.size() + "\n");
        bw.flush();
        bw.close();
    }
}

File: Beakjoon/src/Gold/Lv5/boj_1374.kt
package Gold.Lv5

import java.util.*


data class lecture(
    var index: Int,
    var start: Int,
    var end: Int
)

fun main() { // 강의실
    val N = readln().toInt()
    val arr = Array<lecture?>(N) { null }

    for (i in 0 until N) {
        val inputs = readln().split(" ").map { it.toInt() }
        arr[i] = lecture(inputs[0], inputs[1], inputs[2])
    }

    arr.sortWith(compareBy({ it!!.start }, { it!!.end }))

    val pq = PriorityQueue<Int>()
    for (i in 0 until N) {
        if (pq.isEmpty()) pq.add(arr[i]!!.end)
        else {
            val end_time = pq.peek()
            if (arr[i]!!.start < end_time) pq.add(arr[i]!!.end)
            else {
                pq.poll()
                pq.add(arr[i]!!.end)
            }
        }
    }

    println(pq.size)
}